// element-plus/lib/utils/dom/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_aria = require('./aria.js');
const require_event = require('./event.js');
const require_position = require('./position.js');
const require_style = require('./style.js');
const require_scroll = require('./scroll.js');
const require_element = require('./element.js');

exports.addClass = require_style.addClass;
exports.addUnit = require_style.addUnit;
exports.animateScrollTo = require_scroll.animateScrollTo;
exports.classNameToArray = require_style.classNameToArray;
exports.composeEventHandlers = require_event.composeEventHandlers;
exports.focusElement = require_aria.focusElement;
exports.focusNode = require_aria.focusNode;
exports.getClientXY = require_position.getClientXY;
exports.getElement = require_element.getElement;
exports.getEventCode = require_event.getEventCode;
exports.getEventKey = require_event.getEventKey;
exports.getOffsetTop = require_position.getOffsetTop;
exports.getOffsetTopDistance = require_position.getOffsetTopDistance;
exports.getScrollBarWidth = require_scroll.getScrollBarWidth;
exports.getScrollContainer = require_scroll.getScrollContainer;
exports.getScrollElement = require_scroll.getScrollElement;
exports.getScrollTop = require_scroll.getScrollTop;
exports.getSibling = require_aria.getSibling;
exports.getStyle = require_style.getStyle;
exports.hasClass = require_style.hasClass;
exports.isFocusable = require_aria.isFocusable;
exports.isInContainer = require_position.isInContainer;
exports.isLeaf = require_aria.isLeaf;
exports.isScroll = require_scroll.isScroll;
exports.isShadowRoot = require_aria.isShadowRoot;
exports.isVisible = require_aria.isVisible;
exports.obtainAllFocusableElements = require_aria.obtainAllFocusableElements;
exports.removeClass = require_style.removeClass;
exports.removeStyle = require_style.removeStyle;
exports.scrollIntoView = require_scroll.scrollIntoView;
exports.setStyle = require_style.setStyle;
exports.triggerEvent = require_aria.triggerEvent;
exports.whenMouse = require_event.whenMouse;
