// element-plus/es/locales.mjs
import en_default from "./locale/lang/en.mjs";
import af_default from "./locale/lang/af.mjs";
import ar_eg_default from "./locale/lang/ar-eg.mjs";
import ar_default from "./locale/lang/ar.mjs";
import az_default from "./locale/lang/az.mjs";
import bg_default from "./locale/lang/bg.mjs";
import bn_default from "./locale/lang/bn.mjs";
import ca_default from "./locale/lang/ca.mjs";
import ckb_default from "./locale/lang/ckb.mjs";
import cs_default from "./locale/lang/cs.mjs";
import da_default from "./locale/lang/da.mjs";
import de_default from "./locale/lang/de.mjs";
import el_default from "./locale/lang/el.mjs";
import eo_default from "./locale/lang/eo.mjs";
import es_default from "./locale/lang/es.mjs";
import et_default from "./locale/lang/et.mjs";
import eu_default from "./locale/lang/eu.mjs";
import fa_default from "./locale/lang/fa.mjs";
import fi_default from "./locale/lang/fi.mjs";
import fr_default from "./locale/lang/fr.mjs";
import he_default from "./locale/lang/he.mjs";
import hi_default from "./locale/lang/hi.mjs";
import hr_default from "./locale/lang/hr.mjs";
import hu_default from "./locale/lang/hu.mjs";
import hy_am_default from "./locale/lang/hy-am.mjs";
import id_default from "./locale/lang/id.mjs";
import it_default from "./locale/lang/it.mjs";
import ja_default from "./locale/lang/ja.mjs";
import kk_default from "./locale/lang/kk.mjs";
import km_default from "./locale/lang/km.mjs";
import ko_default from "./locale/lang/ko.mjs";
import ku_default from "./locale/lang/ku.mjs";
import ky_default from "./locale/lang/ky.mjs";
import lo_default from "./locale/lang/lo.mjs";
import lt_default from "./locale/lang/lt.mjs";
import lv_default from "./locale/lang/lv.mjs";
import mg_default from "./locale/lang/mg.mjs";
import mn_default from "./locale/lang/mn.mjs";
import ms_default from "./locale/lang/ms.mjs";
import my_default from "./locale/lang/my.mjs";
import nb_no_default from "./locale/lang/nb-no.mjs";
import nl_default from "./locale/lang/nl.mjs";
import no_default from "./locale/lang/no.mjs";
import pa_default from "./locale/lang/pa.mjs";
import pl_default from "./locale/lang/pl.mjs";
import pt_br_default from "./locale/lang/pt-br.mjs";
import pt_default from "./locale/lang/pt.mjs";
import ro_default from "./locale/lang/ro.mjs";
import ru_default from "./locale/lang/ru.mjs";
import sk_default from "./locale/lang/sk.mjs";
import sl_default from "./locale/lang/sl.mjs";
import sr_default from "./locale/lang/sr.mjs";
import sv_default from "./locale/lang/sv.mjs";
import sw_default from "./locale/lang/sw.mjs";
import ta_default from "./locale/lang/ta.mjs";
import te_default from "./locale/lang/te.mjs";
import th_default from "./locale/lang/th.mjs";
import tk_default from "./locale/lang/tk.mjs";
import tr_default from "./locale/lang/tr.mjs";
import ug_cn_default from "./locale/lang/ug-cn.mjs";
import uk_default from "./locale/lang/uk.mjs";
import uz_uz_default from "./locale/lang/uz-uz.mjs";
import vi_default from "./locale/lang/vi.mjs";
import zh_cn_default from "./locale/lang/zh-cn.mjs";
import zh_tw_default from "./locale/lang/zh-tw.mjs";
import zh_hk_default from "./locale/lang/zh-hk.mjs";
import zh_mo_default from "./locale/lang/zh-mo.mjs";

export { af_default as af, ar_default as ar, ar_eg_default as arEg, az_default as az, bg_default as bg, bn_default as bn, ca_default as ca, ckb_default as ckb, cs_default as cs, da_default as da, de_default as de, el_default as el, en_default as en, eo_default as eo, es_default as es, et_default as et, eu_default as eu, fa_default as fa, fi_default as fi, fr_default as fr, he_default as he, hi_default as hi, hr_default as hr, hu_default as hu, hy_am_default as hyAm, id_default as id, it_default as it, ja_default as ja, kk_default as kk, km_default as km, ko_default as ko, ku_default as ku, ky_default as ky, lo_default as lo, lt_default as lt, lv_default as lv, mg_default as mg, mn_default as mn, ms_default as ms, my_default as my, nb_no_default as nbNo, nl_default as nl, no_default as no, pa_default as pa, pl_default as pl, pt_default as pt, pt_br_default as ptBr, ro_default as ro, ru_default as ru, sk_default as sk, sl_default as sl, sr_default as sr, sv_default as sv, sw_default as sw, ta_default as ta, te_default as te, th_default as th, tk_default as tk, tr_default as tr, ug_cn_default as ugCn, uk_default as uk, uz_uz_default as uzUz, vi_default as vi, zh_cn_default as zhCn, zh_hk_default as zhHk, zh_mo_default as zhMo, zh_tw_default as zhTw };
