// element-plus/lib/utils/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.js');
const require_aria = require('./dom/aria.js');
const require_browser = require('./browser.js');
const require_event = require('./dom/event.js');
const require_position = require('./dom/position.js');
const require_easings = require('./easings.js');
const require_types = require('./types.js');
const require_raf = require('./raf.js');
const require_strings = require('./strings.js');
const require_objects = require('./objects.js');
const require_error = require('./error.js');
const require_style = require('./dom/style.js');
const require_scroll = require('./dom/scroll.js');
const require_element = require('./dom/element.js');
const require_global_node = require('./vue/global-node.js');
const require_runtime$1 = require('./vue/props/runtime.js');
const require_icon = require('./vue/icon.js');
const require_install = require('./vue/install.js');
const require_refs = require('./vue/refs.js');
const require_size = require('./vue/size.js');
const require_validator = require('./vue/validator.js');
const require_vnode = require('./vue/vnode.js');
const require_arrays = require('./arrays.js');
const require_i18n = require('./i18n.js');
const require_rand = require('./rand.js');
const require_typescript = require('./typescript.js');
const require_throttleByRaf = require('./throttleByRaf.js');
const require_numbers = require('./numbers.js');
let _vue_shared = require("@vue/shared");
let _vueuse_core = require("@vueuse/core");
let lodash_unified = require("lodash-unified");

exports.CloseComponents = require_icon.CloseComponents;
Object.defineProperty(exports, 'NOOP', {
  enumerable: true,
  get: function () {
    return _vue_shared.NOOP;
  }
});
exports.PatchFlags = require_vnode.PatchFlags;
exports.TypeComponents = require_icon.TypeComponents;
exports.TypeComponentsMap = require_icon.TypeComponentsMap;
exports.ValidateComponentsMap = require_icon.ValidateComponentsMap;
exports.addClass = require_style.addClass;
exports.addUnit = require_style.addUnit;
exports.animateScrollTo = require_scroll.animateScrollTo;
exports.buildProp = require_runtime$1.buildProp;
exports.buildProps = require_runtime$1.buildProps;
exports.cAF = require_raf.cAF;
exports.camelize = _vue_shared.camelize;
exports.capitalize = require_strings.capitalize;
exports.castArray = require_arrays.castArray;
exports.changeGlobalNodesTarget = require_global_node.changeGlobalNodesTarget;
exports.classNameToArray = require_style.classNameToArray;
exports.composeEventHandlers = require_event.composeEventHandlers;
exports.composeRefs = require_refs.composeRefs;
exports.createGlobalNode = require_global_node.createGlobalNode;
exports.debugWarn = require_error.debugWarn;
exports.definePropType = require_runtime$1.definePropType;
exports.easeInOutCubic = require_easings.easeInOutCubic;
Object.defineProperty(exports, 'ensureArray', {
  enumerable: true,
  get: function () {
    return lodash_unified.castArray;
  }
});
exports.entriesOf = require_objects.entriesOf;
exports.epPropKey = require_runtime$1.epPropKey;
exports.escapeStringRegexp = require_strings.escapeStringRegexp;
exports.extractFirst = require_arrays.extractFirst;
exports.flattedChildren = require_vnode.flattedChildren;
exports.focusElement = require_aria.focusElement;
exports.focusNode = require_aria.focusNode;
exports.generateId = require_rand.generateId;
exports.getClientXY = require_position.getClientXY;
exports.getComponentSize = require_size.getComponentSize;
exports.getElement = require_element.getElement;
exports.getEventCode = require_event.getEventCode;
exports.getEventKey = require_event.getEventKey;
exports.getFirstValidNode = require_vnode.getFirstValidNode;
exports.getNormalizedProps = require_vnode.getNormalizedProps;
exports.getOffsetTop = require_position.getOffsetTop;
exports.getOffsetTopDistance = require_position.getOffsetTopDistance;
exports.getProp = require_objects.getProp;
exports.getRandomInt = require_rand.getRandomInt;
exports.getScrollBarWidth = require_scroll.getScrollBarWidth;
exports.getScrollContainer = require_scroll.getScrollContainer;
exports.getScrollElement = require_scroll.getScrollElement;
exports.getScrollTop = require_scroll.getScrollTop;
exports.getSibling = require_aria.getSibling;
exports.getStyle = require_style.getStyle;
exports.hasClass = require_style.hasClass;
Object.defineProperty(exports, 'hasOwn', {
  enumerable: true,
  get: function () {
    return _vue_shared.hasOwn;
  }
});
exports.hyphenate = _vue_shared.hyphenate;
exports.iconPropType = require_icon.iconPropType;
exports.isAndroid = require_browser.isAndroid;
Object.defineProperty(exports, 'isArray', {
  enumerable: true,
  get: function () {
    return _vue_shared.isArray;
  }
});
exports.isBoolean = require_types.isBoolean;
exports.isClient = _vueuse_core.isClient;
exports.isComment = require_vnode.isComment;
Object.defineProperty(exports, 'isDate', {
  enumerable: true,
  get: function () {
    return _vue_shared.isDate;
  }
});
exports.isElement = require_types.isElement;
exports.isEmpty = require_types.isEmpty;
exports.isEpProp = require_runtime$1.isEpProp;
exports.isFirefox = require_browser.isFirefox;
exports.isFocusable = require_aria.isFocusable;
exports.isFragment = require_vnode.isFragment;
Object.defineProperty(exports, 'isFunction', {
  enumerable: true,
  get: function () {
    return _vue_shared.isFunction;
  }
});
exports.isGreaterThan = require_numbers.isGreaterThan;
exports.isIOS = _vueuse_core.isIOS;
exports.isInContainer = require_position.isInContainer;
exports.isKorean = require_i18n.isKorean;
exports.isLeaf = require_aria.isLeaf;
exports.isNumber = require_types.isNumber;
Object.defineProperty(exports, 'isObject', {
  enumerable: true,
  get: function () {
    return _vue_shared.isObject;
  }
});
Object.defineProperty(exports, 'isPlainObject', {
  enumerable: true,
  get: function () {
    return _vue_shared.isPlainObject;
  }
});
Object.defineProperty(exports, 'isPromise', {
  enumerable: true,
  get: function () {
    return _vue_shared.isPromise;
  }
});
exports.isPropAbsent = require_types.isPropAbsent;
exports.isScroll = require_scroll.isScroll;
exports.isShadowRoot = require_aria.isShadowRoot;
Object.defineProperty(exports, 'isString', {
  enumerable: true,
  get: function () {
    return _vue_shared.isString;
  }
});
exports.isStringNumber = require_types.isStringNumber;
Object.defineProperty(exports, 'isSymbol', {
  enumerable: true,
  get: function () {
    return _vue_shared.isSymbol;
  }
});
exports.isTemplate = require_vnode.isTemplate;
exports.isText = require_vnode.isText;
exports.isUndefined = require_types.isUndefined;
exports.isValidComponentSize = require_validator.isValidComponentSize;
exports.isValidDatePickType = require_validator.isValidDatePickType;
exports.isValidElementNode = require_vnode.isValidElementNode;
exports.isVisible = require_aria.isVisible;
exports.isWindow = require_types.isWindow;
exports.kebabCase = require_strings.kebabCase;
exports.keysOf = require_objects.keysOf;
exports.mutable = require_typescript.mutable;
exports.obtainAllFocusableElements = require_aria.obtainAllFocusableElements;
exports.rAF = require_raf.rAF;
exports.removeClass = require_style.removeClass;
exports.removeGlobalNode = require_global_node.removeGlobalNode;
exports.removeStyle = require_style.removeStyle;
exports.renderBlock = require_vnode.renderBlock;
exports.renderIf = require_vnode.renderIf;
exports.scrollIntoView = require_scroll.scrollIntoView;
exports.setStyle = require_style.setStyle;
exports.throttleByRaf = require_throttleByRaf.throttleByRaf;
exports.throwError = require_error.throwError;
Object.defineProperty(exports, 'toRawType', {
  enumerable: true,
  get: function () {
    return _vue_shared.toRawType;
  }
});
exports.triggerEvent = require_aria.triggerEvent;
exports.unique = require_arrays.unique;
exports.whenMouse = require_event.whenMouse;
exports.withInstall = require_install.withInstall;
exports.withInstallDirective = require_install.withInstallDirective;
exports.withInstallFunction = require_install.withInstallFunction;
exports.withNoopInstall = require_install.withNoopInstall;
exports.withPropsDefaultsSetter = require_install.withPropsDefaultsSetter;
