// element-plus@2.13.7/lib/utils/vue/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_global_node = require('./global-node.js');
const require_runtime = require('./props/runtime.js');
const require_icon = require('./icon.js');
const require_install = require('./install.js');
const require_refs = require('./refs.js');
const require_size = require('./size.js');
const require_validator = require('./validator.js');
const require_vnode = require('./vnode.js');

exports.CloseComponents = require_icon.CloseComponents;
exports.PatchFlags = require_vnode.PatchFlags;
exports.TypeComponents = require_icon.TypeComponents;
exports.TypeComponentsMap = require_icon.TypeComponentsMap;
exports.ValidateComponentsMap = require_icon.ValidateComponentsMap;
exports.buildProp = require_runtime.buildProp;
exports.buildProps = require_runtime.buildProps;
exports.changeGlobalNodesTarget = require_global_node.changeGlobalNodesTarget;
exports.composeRefs = require_refs.composeRefs;
exports.createGlobalNode = require_global_node.createGlobalNode;
exports.definePropType = require_runtime.definePropType;
exports.epPropKey = require_runtime.epPropKey;
exports.flattedChildren = require_vnode.flattedChildren;
exports.getComponentSize = require_size.getComponentSize;
exports.getFirstValidNode = require_vnode.getFirstValidNode;
exports.getNormalizedProps = require_vnode.getNormalizedProps;
exports.iconPropType = require_icon.iconPropType;
exports.isComment = require_vnode.isComment;
exports.isEpProp = require_runtime.isEpProp;
exports.isFragment = require_vnode.isFragment;
exports.isTemplate = require_vnode.isTemplate;
exports.isText = require_vnode.isText;
exports.isValidComponentSize = require_validator.isValidComponentSize;
exports.isValidDatePickType = require_validator.isValidDatePickType;
exports.isValidElementNode = require_vnode.isValidElementNode;
exports.removeGlobalNode = require_global_node.removeGlobalNode;
exports.renderBlock = require_vnode.renderBlock;
exports.renderIf = require_vnode.renderIf;
exports.withInstall = require_install.withInstall;
exports.withInstallDirective = require_install.withInstallDirective;
exports.withInstallFunction = require_install.withInstallFunction;
exports.withNoopInstall = require_install.withNoopInstall;
exports.withPropsDefaultsSetter = require_install.withPropsDefaultsSetter;
