// element-plus@2.13.7/lib/constants/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_aria = require('./aria.js');
const require_column_alignment = require('./column-alignment.js');
const require_date = require('./date.js');
const require_event = require('./event.js');
const require_form = require('./form.js');
const require_key = require('./key.js');
const require_size = require('./size.js');

exports.BORDER_HORIZONTAL_WIDTH = require_form.BORDER_HORIZONTAL_WIDTH;
exports.CHANGE_EVENT = require_event.CHANGE_EVENT;
exports.EVENT_CODE = require_aria.EVENT_CODE;
exports.INPUT_EVENT = require_event.INPUT_EVENT;
exports.INSTALLED_KEY = require_key.INSTALLED_KEY;
exports.MINIMUM_INPUT_WIDTH = require_form.MINIMUM_INPUT_WIDTH;
exports.UPDATE_MODEL_EVENT = require_event.UPDATE_MODEL_EVENT;
exports.WEEK_DAYS = require_date.WEEK_DAYS;
exports.columnAlignment = require_column_alignment.columnAlignment;
exports.componentSizeMap = require_size.componentSizeMap;
exports.componentSizes = require_size.componentSizes;
exports.datePickTypes = require_date.datePickTypes;
