// element-plus@2.13.7/lib/components/empty/src/img-empty.vue_vue_type_script_setup_true_lang.js
const require_runtime = require('../../../_virtual/_rolldown/runtime.js');
const require_index = require('../../../hooks/use-namespace/index.js');
const require_index$1 = require('../../../hooks/use-id/index.js');
let vue = require("vue");

//#region ../../packages/components/empty/src/img-empty.vue?vue&type=script&setup=true&lang.ts
const _hoisted_1 = {
	viewBox: "0 0 79 86",
	version: "1.1",
	xmlns: "http://www.w3.org/2000/svg",
	"xmlns:xlink": "http://www.w3.org/1999/xlink"
};
const _hoisted_2 = ["id"];
const _hoisted_3 = ["stop-color"];
const _hoisted_4 = ["stop-color"];
const _hoisted_5 = ["id"];
const _hoisted_6 = ["stop-color"];
const _hoisted_7 = ["stop-color"];
const _hoisted_8 = ["id"];
const _hoisted_9 = {
	stroke: "none",
	"stroke-width": "1",
	fill: "none",
	"fill-rule": "evenodd"
};
const _hoisted_10 = { transform: "translate(-1268.000000, -535.000000)" };
const _hoisted_11 = { transform: "translate(1268.000000, 535.000000)" };
const _hoisted_12 = ["fill"];
const _hoisted_13 = ["fill"];
const _hoisted_14 = { transform: "translate(34.500000, 31.500000) scale(-1, 1) rotate(-25.000000) translate(-34.500000, -31.500000) translate(7.000000, 10.000000)" };
const _hoisted_15 = ["fill"];
const _hoisted_16 = ["fill"];
const _hoisted_17 = ["fill"];
const _hoisted_18 = ["fill"];
const _hoisted_19 = ["fill"];
const _hoisted_20 = { transform: "translate(53.000000, 45.000000)" };
const _hoisted_21 = ["fill", "xlink:href"];
const _hoisted_22 = ["fill", "mask"];
const _hoisted_23 = ["fill"];
var img_empty_vue_vue_type_script_setup_true_lang_default = /* @__PURE__ */ (0, vue.defineComponent)({
	name: "ImgEmpty",
	__name: "img-empty",
	setup(__props) {
		const ns = require_index.useNamespace("empty");
		const id = require_index$1.useId();
		return (_ctx, _cache) => {
			return (0, vue.openBlock)(), (0, vue.createElementBlock)("svg", _hoisted_1, [(0, vue.createElementVNode)("defs", null, [
				(0, vue.createElementVNode)("linearGradient", {
					id: `linearGradient-1-${(0, vue.unref)(id)}`,
					x1: "38.8503086%",
					y1: "0%",
					x2: "61.1496914%",
					y2: "100%"
				}, [(0, vue.createElementVNode)("stop", {
					"stop-color": `var(${(0, vue.unref)(ns).cssVarBlockName("fill-color-1")})`,
					offset: "0%"
				}, null, 8, _hoisted_3), (0, vue.createElementVNode)("stop", {
					"stop-color": `var(${(0, vue.unref)(ns).cssVarBlockName("fill-color-4")})`,
					offset: "100%"
				}, null, 8, _hoisted_4)], 8, _hoisted_2),
				(0, vue.createElementVNode)("linearGradient", {
					id: `linearGradient-2-${(0, vue.unref)(id)}`,
					x1: "0%",
					y1: "9.5%",
					x2: "100%",
					y2: "90.5%"
				}, [(0, vue.createElementVNode)("stop", {
					"stop-color": `var(${(0, vue.unref)(ns).cssVarBlockName("fill-color-1")})`,
					offset: "0%"
				}, null, 8, _hoisted_6), (0, vue.createElementVNode)("stop", {
					"stop-color": `var(${(0, vue.unref)(ns).cssVarBlockName("fill-color-6")})`,
					offset: "100%"
				}, null, 8, _hoisted_7)], 8, _hoisted_5),
				(0, vue.createElementVNode)("rect", {
					id: `path-3-${(0, vue.unref)(id)}`,
					x: "0",
					y: "0",
					width: "17",
					height: "36"
				}, null, 8, _hoisted_8)
			]), (0, vue.createElementVNode)("g", _hoisted_9, [(0, vue.createElementVNode)("g", _hoisted_10, [(0, vue.createElementVNode)("g", _hoisted_11, [
				(0, vue.createElementVNode)("path", {
					d: "M39.5,86 C61.3152476,86 79,83.9106622 79,81.3333333 C79,78.7560045 57.3152476,78 35.5,78 C13.6847524,78 0,78.7560045 0,81.3333333 C0,83.9106622 17.6847524,86 39.5,86 Z",
					fill: `var(${(0, vue.unref)(ns).cssVarBlockName("fill-color-3")})`
				}, null, 8, _hoisted_12),
				(0, vue.createElementVNode)("polygon", {
					fill: `var(${(0, vue.unref)(ns).cssVarBlockName("fill-color-7")})`,
					transform: "translate(27.500000, 51.500000) scale(1, -1) translate(-27.500000, -51.500000) ",
					points: "13 58 53 58 42 45 2 45"
				}, null, 8, _hoisted_13),
				(0, vue.createElementVNode)("g", _hoisted_14, [
					(0, vue.createElementVNode)("polygon", {
						fill: `var(${(0, vue.unref)(ns).cssVarBlockName("fill-color-7")})`,
						transform: "translate(11.500000, 5.000000) scale(1, -1) translate(-11.500000, -5.000000) ",
						points: "2.84078316e-14 3 18 3 23 7 5 7"
					}, null, 8, _hoisted_15),
					(0, vue.createElementVNode)("polygon", {
						fill: `var(${(0, vue.unref)(ns).cssVarBlockName("fill-color-5")})`,
						points: "-3.69149156e-15 7 38 7 38 43 -3.69149156e-15 43"
					}, null, 8, _hoisted_16),
					(0, vue.createElementVNode)("rect", {
						fill: `url(#linearGradient-1-${(0, vue.unref)(id)})`,
						transform: "translate(46.500000, 25.000000) scale(-1, 1) translate(-46.500000, -25.000000) ",
						x: "38",
						y: "7",
						width: "17",
						height: "36"
					}, null, 8, _hoisted_17),
					(0, vue.createElementVNode)("polygon", {
						fill: `var(${(0, vue.unref)(ns).cssVarBlockName("fill-color-2")})`,
						transform: "translate(39.500000, 3.500000) scale(-1, 1) translate(-39.500000, -3.500000) ",
						points: "24 7 41 7 55 -3.63806207e-12 38 -3.63806207e-12"
					}, null, 8, _hoisted_18)
				]),
				(0, vue.createElementVNode)("rect", {
					fill: `url(#linearGradient-2-${(0, vue.unref)(id)})`,
					x: "13",
					y: "45",
					width: "40",
					height: "36"
				}, null, 8, _hoisted_19),
				(0, vue.createElementVNode)("g", _hoisted_20, [(0, vue.createElementVNode)("use", {
					fill: `var(${(0, vue.unref)(ns).cssVarBlockName("fill-color-8")})`,
					transform: "translate(8.500000, 18.000000) scale(-1, 1) translate(-8.500000, -18.000000) ",
					"xlink:href": `#path-3-${(0, vue.unref)(id)}`
				}, null, 8, _hoisted_21), (0, vue.createElementVNode)("polygon", {
					fill: `var(${(0, vue.unref)(ns).cssVarBlockName("fill-color-9")})`,
					mask: `url(#mask-4-${(0, vue.unref)(id)})`,
					transform: "translate(12.000000, 9.000000) scale(-1, 1) translate(-12.000000, -9.000000) ",
					points: "7 0 24 0 20 18 7 16.5"
				}, null, 8, _hoisted_22)]),
				(0, vue.createElementVNode)("polygon", {
					fill: `var(${(0, vue.unref)(ns).cssVarBlockName("fill-color-2")})`,
					transform: "translate(66.000000, 51.500000) scale(-1, 1) translate(-66.000000, -51.500000) ",
					points: "62 45 79 45 70 58 53 58"
				}, null, 8, _hoisted_23)
			])])])]);
		};
	}
});

//#endregion
exports.default = img_empty_vue_vue_type_script_setup_true_lang_default;
