// element-plus@2.13.7/es/components/roving-focus-group/index.mjs
import { ROVING_FOCUS_COLLECTION_INJECTION_KEY as COLLECTION_INJECTION_KEY, ROVING_FOCUS_ITEM_COLLECTION_INJECTION_KEY as COLLECTION_ITEM_INJECTION_KEY } from "./src/roving-focus-group.mjs";
import { ROVING_FOCUS_GROUP_INJECTION_KEY, ROVING_FOCUS_GROUP_ITEM_INJECTION_KEY } from "./src/tokens.mjs";
import { focusFirst, getFocusIntent, reorderArray } from "./src/utils.mjs";
import roving_focus_group_default from "./src/roving-focus-group2.mjs";
import roving_focus_item_default from "./src/roving-focus-item.mjs";

//#region ../../packages/components/roving-focus-group/index.ts
var roving_focus_group_default$1 = roving_focus_group_default;

//#endregion
export { roving_focus_group_default as ElRovingFocusGroup, roving_focus_item_default as ElRovingFocusItem, COLLECTION_INJECTION_KEY as ROVING_FOCUS_COLLECTION_INJECTION_KEY, ROVING_FOCUS_GROUP_INJECTION_KEY, ROVING_FOCUS_GROUP_ITEM_INJECTION_KEY, COLLECTION_ITEM_INJECTION_KEY as ROVING_FOCUS_ITEM_COLLECTION_INJECTION_KEY, roving_focus_group_default$1 as default, focusFirst, getFocusIntent, reorderArray };
